### DISCUSSION TITLE
Task.FromResult() can not infer nullable types directly

### DISCUSSION THREAD

**Discussion by marwalsch** [dotnet/csharplang#5127] [Q&A] [answered]
I hope this is the right place for this type question.

Assume the following method signature:

`public Task<Guid?> Foo()`

Let some inheritance or interface compliance require this method to return `Task` but lack the need to be `async`.
Given that, the following can be returned without error:

```
Guid? emtpyId = Guid.Empty;
return Task.FromResult(emtpyId);
```
However, returning `Task.FromResult(Guid.Empty)` does not work as Guid.Empty is not nullable. The same applies to `Task.FromResult(null)` 

Why is that? Shouldn't the compiler be aware that assigning a `Guid` or `null` to a `Guid?` are valid operations?

**Comment by Joe4evr** [comment, marked as answer]
> Why is that? Shouldn't the compiler be aware that assigning a `Guid` or `null` to a `Guid?` are valid operations?

No it does not. For one thing, `Guid` and `Guid?` are completely different types, and type arguments on classes and structs must be exact because they can't be co-/contravariant. Even when using reference types, [you can't convert a `Task<string>` to a `Task<object>`](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEA3GUCWAZgJ5oAmIA1AD4ACATAIwCwAULQAwAEtjArAG42bBlwAKbAN5susngGYeSADwRgAKxhgMAPi4BZABQBKGXOms5VngHYlAOgBiUCAFsASjADOAVwA2GIYARAASMH5+EFwA7tB+ZEHGQpZyAL5sqUA=), and you can't convert a `Task<Guid>` to a `Task<Guid?>` for the same reason.
On the other hand, the statement `Guid? emtpyId = Guid.Empty;` simply uses the implicit conversions defined on `Nullable<T>` (which can't be used for inferring type arguments) and so doesn't produce the same error.

Note, by the way, that [a similar restriction holds with Nullable Reference Types as well](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEA3GUCWAZgJ5oAmIA1AD4ACATAIwCwAULQAwAEtjArAG42bAMQA7AK4AbKQENgUmFxhj5itgy4AFNgG82XQzwDMPJAB5eHAPwA+LgFkAFAEoDR/ayPeeAdjMAdABiUBAAtgBKMADO0hhOAEQAEjAyEFwA7tBSZAEJLkJeRgC+bMVAA==), albeit with a warning rather than an error. This is one thing I'm hoping can be addressed at some point so generic inference and NRTs can play together better, but it's probably not a high priority.

**Reply by marwalsch** [reply]
I wasn't aware of the strict distinction between RTs and NRTs. Thanks for elaborating.

**Comment by alrz** [comment]
This relates to https://github.com/dotnet/csharplang/discussions/92

`return Task.FromResult(null)` could be resolved as `Task.FromResult<Guid?>(null)` based on the target-type, however for `FromResult(Guid.Empty)` it won't work because the type inference has already succeeded with `T=Guid`.

**Reply by marwalsch** [reply]
`Task.FromResult<Guid?>(null)` is what I was looking for. I initially stumbled over the issue as I simply returned null directly thinking it would infer to Task<null>, which it doesn't (which wakes sense looking at it now).

[Related Issue](https://github.com/JosefPihrt/Roslynator/issues/180) for any Googlers.
